### PR TITLE
[WIP] Add Service API To Snippets

### PR DESCRIPTION
**Pull request by joefitzgerald**
:rotating_light: WIP :rotating_light:

Not ready for merge yet:

- [x] Add API
- [ ] Add tests
- [ ] Update autocomplete-plus to consume the API and validate it has the right surface area

:rotating_light: WIP :rotating_light:

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,5 +18,13 @@
     "season": "^5.0.4",
     "temp": "~0.8.0",
     "underscore-plus": "^1.0.0"
+  },
+  "providedServices": {
+    "snippets": {
+      "description": "Snippets are text shortcuts that can be expanded to their definition.",
+      "versions": {
+        "0.73.0": "provideSnippets"
```
Comment (benogle) [inline package.json]: This should probably be 1.0.0

File: lib/snippets.coffee [hunk 2]
```diff
@@ -273,9 +273,18 @@ module.exports =
...
     new SnippetExpansion(snippet, editor, cursor, this)
+
+  provideSnippets: ->
+    isLoaded = -> @loaded
+    snippets = this
+    return s =
+      loaded: isLoaded.bind(snippets)
+      insert: @insert.bind(snippets)
+      getSnippets: @getSnippets.bind(snippets)
+      getPrefixText: @getPrefixText.bind(snippets)
```
Comment (benogle) [inline lib/snippets.coffee]: I think we should leave this super minimal and just provide the `insert` method. It's really all autocomplete needs to do its thing.

File: package.json (hunk 1 above)
Comment (joefitzgerald) [inline package.json]: Was just following the pattern @maxbrunsfeld seemed to establish in status-bar.

File: lib/snippets.coffee (hunk 2 above)
Comment (joefitzgerald) [inline lib/snippets.coffee]: These additional things cater to `autocomplete-snippets` needs. And isLoaded is required to allow testing to be possible.

File: package.json (hunk 1 above)
Comment (benogle) [inline package.json]: status-bar wasnt the ideal scenario. We want to version the service independently from the package version.

Comment (joefitzgerald) [inline package.json]: :thumbsup:

Comment (benogle) [inline package.json]: After talking to @maxbrunsfeld we decided this should probably start at `0.1.0` rather than `1.0.0`.

Comment (joefitzgerald) [inline package.json]: Sounds good.

File: lib/snippets.coffee (hunk 2 above)
Comment (benogle) [inline lib/snippets.coffee]: autocomplete-snippets only uses `loaded` in specs and it get the snippets via `atom.config`. Getting the snippets from config provides less surface area and is much more flexible than the `getSnippets` method. Looks like `getPrefixText` is not used anywhere.

Maybe loaded stays, but with a new name for more clarity `allSnippetsLoaded` or `bundledSnippetsLoaded`.

**Comment by benogle** [comment]
I can take this over as I am tinkering with snippets in autocomplete plus

File: lib/snippets.coffee (hunk 2 above)
Comment (joefitzgerald) [inline lib/snippets.coffee]: getPrefixText would be used to replace the default prefix that is provided by autocomplete-plus in the suggestion request: https://github.com/atom-community/autocomplete-snippets/blob/7e183365030cfa96586077d0e7eb339d5b52eace/lib/snippets-provider.coffee#L9. I would assume the snippets package has better logic for determining the prefix of a snippet than we do, but perhaps this is moot given the simplicity of most prefixes (they are generally `[\w]*`).

Comment (benogle) [inline lib/snippets.coffee]: I would rather the prefix were generated from the autocomplete package or by asking the providers.

**Pull request closed without merging**